### ISSUE TITLE
JDK9+ Jigsaw modules support

### ISSUE THREAD

**Issue by oli0044** [graphhopper/graphhopper#2053]
Hi,

I'm trying to add the maven dependency graphhopper-reader-osm as suggested in the docs to a JDK 11 Java project using the new Jigsaw module system (i.e. module-info.java exists) ... [, because the Jigsaw module system is needed by other dependencies].

Up to graphhopper version 0.10.3 it works peachy, but since version 0.11.0 up to the current 1.0 the project does not compile anymore with the error:
`reads package com.graphhopper from both graphhopper.api and graphhopper.core`

This is due to the fact, that since 0.11.0 some classes from graphhopper-core were split off into graphhopper-api, but the new Jigsaw module system needs the different modules/dependency-jars to be disjoint package-wise. So for example: com.graphhopper.util has some classes in maven dependency graphhopper-ore and some in graphhopper-api, but this is not supported anymore when using the Jigsaw module system since JDK9+ because it needs to have com.graphhopper.util only be exported by one module/dependency.

 - Is there a plan to support the Jigsaw module system in the future with embedded module-info.java files?
 - Is it possible to make a third (optional) maven dependency/module "graphhopper-core-plus-api" which combines graphhopper-core and graphhopper-apiback into one dependency again? (Which - I think - would solve the problem above)
 - Or am I just being dumb and doin something wrong... and the support for the Jigsaw module system is working already all along?

I might maybe be able to suggest a pull request with the second option "combined graphhopper-core-plus-api". But as an only intermediate maven user I would highly appreciate any push in the right direction, or maybe a "stop" if this is the wrong direction.

Thanks for the great work with Graphhopper in any case :-)

Greetings from Germany

Oli4

**Comment by easbar** [comment]
If jigsaw standard does not allow same package names in different modules I think we should not do it. Especially (core) `com.graphhopper.util` is quite a mess anyway (it has many unrelated classes) and should be easy to rename. `com.graphhopper` only has a few classes so they could be moved to some sub-package? Are these the only two packages that cause a problem?

**Comment by oli0044** [comment]
Hi easbar,

thanks for taking this seriously :-).

As far as I see it, It's the packages:

`com.graphhopper`
`com.graphhopper.col`
`com.graphhopper.routing.util`
`com.graphhopper.util`

`com.graphhopper` is especially tricky, because e.g. Graphhopper and GraphhopperAPI are understandebly in core and api respecitvely but should be separated into different subpackages to adhere to the Jigsaw standard. My understanding is, that Jigsaw would like to have
`com.graphhopper.core`
`com.graphhopper.api`
`com.graphhopper.web`
etc.

As for the asked for reference, the stackflow answer https://stackoverflow.com/questions/51828014/how-split-packages-are-avoided-in-java-9 better explains the official jdk9 spec description http://openjdk.java.net/projects/jigsaw/spec/sotms/#readability.

If you would like to move the classes, my suggestion would be to move the split packages
`com.graphhopper.col`
`com.graphhopper.routing.util`
`com.graphhopper.util`
from the core module --> into the api module, so that the packages are only in one module again. Since the core module already depends on the api module, nothing should break...

Then it would suffice to put the classes under `com.graphhopper` in the module api into a subpackage `com.graphhopper.api` or `com.graphhopper.xyz` as you suggested and everything should be sorted and separated again for Jigsaw.

Would you be more ok with that solution?

**Comment by oli0044** [comment]
What I'm just realising:
It could also be possible to just move the Graphhopper.java from core into api, then the `com.graphhopper` package would not be split anymore (so no need for `com.graphhopper.api` then)

**Comment by oli0044** [comment]
OK, I tried the "moving" idea... unfortenately due to massive cross-references in the classes/packages, I ended up having to move everything from core module into api module... letting the core module to have no classes anymore.

I doubt that this is a solution you would like to have, is it ;-)?

I was very careful, not to move anything unnecessary, but unfortenately with every move of a class there were new cross-references that resulted in moving more and more classes.

If you would like to check the commit, in case this **is** an acceptable solution: https://github.com/oli0044/graphhopper/tree/issue_2053b
... but I really doubt it. I'm even too ashamed to transform it into a pull request ;-).

With the idea of introducing a package combining core and api called `coreapi` or `core-plus-api`, the separation of core and api would not be lost and for the time necessary to separate the modules properly, the JDK9+ project developers would have a chance to use the versions higher than 0.10.3.

**Comment by easbar** [comment]
I thought maybe there is no need to move any code between the modules, but we can just use different package names as for example `com.graphhopper.util` in core is not necessarily related to the package with the same name in api. Using a prefix like `.api` or `.core` could be an option. But lets see what others have to say.

**Comment by oli0044** [comment]
> I thought maybe there is no need to move any code between the modules, but we can just use different package names as for example `com.graphhopper.util` in core is not necessarily related to the package with the same name in api (and package privacy isn't used very often in GH either). 

yes, that would be the "good" solution intended by the jigsaw standard, to introduce com.graphhopper.api.util and com.graphhopper.core.util.

> Using a prefix like `.api` or `.core` for the package names could be an option. But lets see what others have to say.

Ok

**Comment by oli0044** [comment]
FYI, I tried your suggestion and moved the classes in the api module into a subpackage `com.graphhopper.api`. Seems to work, so if this will be accepted by the others, I could provide a pull request.

**Comment by michaz** [comment]
I fully support the package-renaming solution.

**Comment by oli0044** [comment]
Just to be completely transparent:

After I moved the content of the api module into a package `com.graphhopper.api` and fixed all the imports (using Netbeans 11.3) there was one Test that failed in the core module:
`com.graphhopper.util.DistanceCalcEuclideanTest`

This was due to the function `calcShrinkFactor()` in `com.graphhopper.api.util.DistanceCalcEarth` which had no private/protected/public declaration and therefore was only visible to same-package classes in `com.graphhopper.api.util`. However `com.graphhopper.util.DistanceCalcEuclidean` now was a child class, but not in the same package anymore which led to the wrong function to be called during runtime. It could be fixed easily by just declaring both funtions `calcShrinkFactor()` as `protected`, so that visibility was also ensured to the child classes.

However that made me think, that there might be some more occurences of methods that are only visible in the same package and therefore not called properly anymore. It turned out, that there were only 3 other occurences of such functions that had no public/protected/public declaration and all 3 functions were only used in the same package.

So I think no worries in that context.

However I just wanted to be careful, complete and transparent... maybe this sparks an idea for anyone, which should be considered...

(I'll attach my documentation for the search of all the occurences of non-public/private/procted-methods)
[graphhopper.functions.txt](https://github.com/graphhopper/graphhopper/files/4753624/graphhopper.functions.txt)

**Comment by oli0044** [comment]
For which branch should I propose the pull request?

`1.0` or `master`?

**Comment by michaz** [comment]
(And I also think this not only solves the immediate problem, but is the totally correct thing to do. Thanks for taking this on!)

**Comment by oli0044** [comment]
np, that's one of the beauties of open source. Thx for the great tool and work :-).

ok, I proposed the pull request. At the moment I purposefully only added package `com.graphhopper.api`, not to change tooo much and also not knowing what future structure you would prefer.

Would you like me to introduce another subpackage already or is it ok this way for the beginning?

**Comment by easbar** [comment]
Can we do all the changes needed prior to actually renaming the package first / in a separate PR? The goal should be that we can update downstream dependencies or other branches by a single 'rename package' refactoring using the IDE. Otherwise it will be a pain to update code that is not in master yet or just uses the code in this repository.

**Comment by oli0044** [comment]
> Can we do all the changes needed prior to actually renaming the package first / in a separate PR? The goal should be that we can update downstream dependencies or other branches by straight forward 'rename package' or 'move files' refactorings using the IDE. Otherwise it will be a pain to update code that is not in master yet or just uses the code in this repository.

I think, I understand what you mean and provided a PR with the changes before moving into the submodule.

Refactoring with the IDE (I used Netbeans) already took care of 90% of the job when moving the files into the subpackage. The remaining 10% was for classes that used wildcard-imports like `com.graphhopper.util.*`. For those Netbeans (at least) does not know what to do when refactoring and I had to fix the imports "manually" (by opening the classes and pressing Ctrl+Shift+I). 

And in one class there was the issue, that `protected` fields were used, instead of pre-existing `public` getter-methods. What's your take on this? Should the `protected` fields made `public` in the new package `com.graphhopper.api`, so that no depending code is broken? Or should the developers be "pushed" to use the already existing `public` getters and setters?

All in all I think what you would like to achieve is already done in the first PR. 90% is done automatically by the IDE when renaming the package. The 10% manual rework, I do not see, how to minimize it further.

> Ok, but you only considered this repository right? For example there is also GH map-matching and navigation. If there are methods that have the same name and signature and do something different we should rename them before putting them into the same package or make them visible outside of some package.

Indeed, I did not know about the other repositories, but a quick look showed, that they are already in separate sub-packages `com.graphhopper.navigation` and `com.graphhopper.matching`. So there is no issue with them.

The only problem arises if a repository/downstream dependency etc. is in the same package as one of the moved subpackages inside the new `com.graphhopper.api` (e.g. `com.graphhopper.util` which now becomes partly `com.graphhopper.api.util`). So I believe only the core module might be in fact affected and thus only repositories and dependencies in combination with the core module.

**Comment by easbar** [comment]
Cool thanks for this!

> The remaining 10% was for classes that used wildcard-imports like com.graphhopper.util.*

I find these rather annoying anyway. I think the default IntelliJ settings we use make use of star imports as soon as there are five or more imports from the same package. But this also means that sometimes the import statements go back and forth when the number of imports changes. Maybe we should set this limit to zero (never use star import) and apply this to the whole code base before renaming the package(s).

> And in one class there was the issue, that protected fields were used, instead of pre-existing public getter-methods. What's your take on this? Should the protected fields made public in the new package com.graphhopper.api, so that no depending code is broken? Or should the developers be "pushed" to use the already existing public getters and setters?

Hm hard to say in general I did not look at these cases yet.

> Indeed, I did not know about the other repositories, but a quick look showed, that they are already in separate sub-packages com.graphhopper.navigation and com.graphhopper.matching. So there is no issue with them.

I was more concerned about other repositories or branches making use of the classes we are moving to another package, because this requires updating all the import statements.

**Comment by easbar** [comment]
And would it not be much easier (to achieve the goal of JDK9+ compatibility) to move the classes of the *other* (not graphhopper-api which is used most frequently) modules first? 

Also before doing this maybe we should have another try of applying the auto-formatting to the whole code-base, because this could also become messy with other branches otherwise, see this discussion: https://github.com/graphhopper/graphhopper/pull/1638#issuecomment-574628308

**Comment by oli0044** [comment]
> I was more concerned about other repositories or branches making use of the classes we are moving to another package, because this requires updating all the import statements

Yes, unfortenately this seems to have to be the case... but at least I believe we think alike: "Do not change anything more than you absolutely have to... ;-)"

> Hm, ok, I see `core/com.graphhopper.util` is also used very often...

Yes exactely. The JDK9+ Jigsaw compatibility issues seem to mainly be only between api and core (since they share the split packages)... and core should even be used more often in dependencies... I'd imagine.

**Comment by oli0044** [comment]
Is there a verdict on how to proceed? Can you foresee a (non-binding) approximate timeline for when this could be included in a release?

Depending on this, I'd be able to include version 1.0 in our project instead of 0.10.3

Greetings  from Germany.

**Comment by michaz** [comment]
The (preliminary) verdict seems to be that we first go forward with trying to use _fewer_ classes in the `api` module. And then package-rename the rest.

 Might take a while, it's a bit hairy, there's a reason why we haven't done it yet.

**Comment by oli0044** [comment]
Hi michaz,

thanks for the quick feedback. Yes I understand, that will take a while for 402 classes ;-).

Could you then please re-consider my proposal of introducing a further maven-package `graphhopper-coreapi` as suggested in PR #2055 (https://github.com/graphhopper/graphhopper/pull/2055)?
With this proposal no change to the code would be necessary, but the JDK-9+ developers would not be shut out bit by bit (The problem is, that as soon as a maven project has another dependency which introduces a modules-info.java file for the dependency, the problem with the split classes in graphhopper-core and -api will prevent the project from compiling. So my fear is, that as the other maven dependencies slowly but surely will be updated to have a modules-info.java, graphhopper will become less and less compatible to projects).

Thanks for your consideration.

Greetings from Germany

**Comment by michaz** [comment]
Since it's just a temporary workaround anyway - can't you do that on your side instead?

I don't think I want to do this. I'm not an expert for the new module system, so I don't really want to publish a hack for which I don't fully understand its implications.

**Comment by michaz** [comment]
I will go forward with orthogonalizing the packages though.

**Comment by oli0044** [comment]
Well, for using it for our project, I would have to publish the maven dependency to a public maven repository. I think I'd call it `graphhopper-coreapi` in my own namespace.

Would I get your permission to do this? I wouldn't change anything in the code, just combine both modules into one and compile it... and publish it like described here: http://tutorials.jenkov.com/maven/publish-to-central-maven-repository.html

I'd keep the apache 2.0 license as used by yourself as well as the other parameters in the current pom-file, just using a sperate namespace.

So then maybe even other JDK9+ developers could use this workaround.

Would this be ok for you? Would you give me the permission to do this?

**Comment by oli0044** [comment]
Hi,

any news here? 

Would you allow for me to compile the preexisting source code of 1.0 with identical license and author recognisation into a new module graphhopper-coreapi and publish it onto a public maven repository?

Would be great, because that would allow us to use graphhopper 1.0 as well as maybe some jdk9+ developers who would be able to use this as a workaround, until the packages are orthogonalized.

Greetings from Germany

**Comment by karussell** [comment]
@oli0044 Or can you use `gh-coreapi` for simplicity :) ?

**Comment by easbar** [comment]
@oli0044 I just removed the api module in #2372. It's contents are now in web-api. Maybe this also helps for this issue? 

There are still some packages split over multiple modules:

* core and web-api both use `com.graphhopper`
* core and web-api both use `com.graphhopper.util`
* web and web-bundle both use `com.graphhopper.http`

And can we close #2066 and #2064? I think they are no longer up to date, especially because of #2372.

**Comment by oli0044** [comment]
Hello @easbar 

The changes in the current version 4.0 are definitely going in the right direction concerning this #2053.

For my project to work with the newest Graphhopper version again it would suffice to move the classes in `com.graphhopper` from the web-api module into the core module as well as the classes in `com.graphhopper.util` from web-api into core.

I believe these classes also belong more into the core module than into the web-api module or am I wrong?

I'll try to suggest a renewed pull request that would do that later this evening.

#2066 and #2064 can be closed then, because they are not up to date anymore.

Greetings

**Comment by easbar** [comment]
> I believe these classes also belong more into the core module than into the web-api module or am I wrong?

I think they are in web-api, because they are needed for the `client-hc` module, our hand-crafted(=hc) Java client. The idea was probably that it should be as lightweight as possible, i.e. not depend on core. For Jigsaw compatibility wouldn't it be sufficient if all modules used their own package names (instead of spreading packages like `com.graphhopper` over multiple modules?

**Comment by oli0044** [comment]
yes indeed, that would be sufficient and imho the most durable solution.

However giving each module its own subpackage - which would be the cleanest and most durable solution - would of course lead into massive refactoring of all import lines (also in dependent projects).

A not so intrusive solution might be to move `com.graphhopper` inside web-api into `com.graphhopper.api` and `com.graphhopper.util` inside web-api into `com.graphhopper.api.util`.

Might that be a compromise? Then the imports won't be changed aaaas massive as if each module gets its own subpackage.

How do you want to solve it? I can offer to create a pull request with the chosen solution, because I am very interested in using the new versions of graphhopper again.

**Comment by easbar** [comment]
> However giving each module its own subpackage - which would be the cleanest and most durable solution - would of course lead into massive refactoring of all import lines (also in dependent projects).

Yes, I'd rather change the package names only for the classes of packages that are currently distributed over multiple modules, i.e. minimal changes to fulfill the Jigsaw compatibility requirement for now. GraphHopper's package structure needs an overhaul anyway, so if we change the package names we should actually think about what packages we want first, but that would be mostly for another issue.

Let's see what we got: 

`com.graphhopper`:

core: `GraphHopper`,`GraphHopperConfig`
web-api: `GHRequest`,`GHResponse`,`Trip`,`ResponsePath`

`com.graphhopper.util`:

core: `AngleCalc`,`ArrayUtil`,`BitUtil/Big/Little`,`Breadth/Depth/XFirstSearch`,`DistanceCalc/3D/Earth/Euclidean/Projection`,`Constants`,`DouglasPeucker`,`Downloader`,`EdgeExplorer/Iterator/State`,`FetchMode`,`GHUtility`,`GitInfo`,`Instructions`,`MiniPerfTest`,`NotThreadSafe`,`PathMerger`,`PathSimplification`,`ProgressListener`,`SimpleIntDeque`,`StopWatch`,`TranslationMap`,`Unzipper` (wow this one is crazy!)

web-api: `Custommodel`,`FinishInstruction`,`Helper`,`Instruction`,`InstructionList`,`JsonFeature`,`JsonFeatureCollection`,`NumHelper`,`Parameters`,`PMap`,`PointAccess`,`PointList`,`RoundaboutInstruction`,`ShallowImmutablePointList`,`Translation`,`ViaInstruction`

`com.graphhopper.http`:

web: `CORSFilter`, `GraphHopperApplication`, `GraphHopperServerConfiguration`
web-bundle: `DurationParam`,`FeedConfiguration`,`GHJerseyViolationExceptionMapper`,`GHLocationParam`,`GHPointParam`,`GraphHopper/RealtimeBundle`,`GraphHopper/RealtimeBundleConfiguration`,`GraphHopperManaged`,`IllegalArgument/MultiExceptionMapper`,`JsonErrorEntity`,`MultiExceptionGPXMessageBodyWriter`,`PtRedirectFilter`,`RealtimeConfiguration`,`RealtimeFeedLoadingCache`,`TypeGPXFilter`

I would assume that the higher-level classes are used in downstream code the most. It is probably least invasive to leave the web-api module untouched. So we need to move `GraphHopper/Config` in core to another package (I think that's ok, and definitely better than changing the package name for `GHRequest/Response` for example).

Most of the classes in core's `com.graphhopper.util` are rather low level and it should be ok to move them to another package. Especially because 'util` does not seem to be a create package for them to be in anyway. I wonder if some of them should go to the `web-api` module. @karussell I can't find it anymore, but wasn't there an issue that was about moving distance calc stuff (or instructions code(?)) into client-hc? 

`com.graphhopper.http` should be easy to fix, because these classes are probably hardly ever used explicitly in downstream code. Maybe we can move the three classes that are in `web` into `web-bundle`?

>  I can offer to create a pull request with the chosen solution, because I am very interested in using the new versions of graphhopper again.

Thanks, that is very nice, I am not sure if this really will speed up the process, also because if we do this refactoring I will probably do it for our own downstream dependencies as well, so I can just as well do this. First we need to decide where the different classes should go though. @karussell, @michaz any thoughts?

**Comment by easbar** [comment]
And of course the easiest thing we can do is deleting classes, instead of moving them. `NotThreadSafe` and `ProgressListener` are very high on my list. The first one we don't really use consistently (only in `RoutingAlgorithm`), and the second one is just a `LongConsumer`. `Instructions` is only used in tests, so maybe we can remove it as well?

**Comment by karussell** [comment]
> @karussell I can't find it anymore, but wasn't there an issue that was about moving distance calc stuff (or instructions code(?)) into client-hc?

Maybe you meant this? https://github.com/graphhopper/graphhopper/issues/2250

**Comment by easbar** [comment]
> Maybe you meant this? #2250

Ah, yes exactly :) So we should fix this one first.

**Comment by oli0044** [comment]
Hi,

long time no read ;-).

I saw that there were a lot of steps into the right direction to orthogonalize the modules.

However there are still the issues of the classes in

`com.graphhopper`:

core: `GraphHopper`,`GraphHopperConfig`
web-api: `GHRequest`,`GHResponse`,`Trip`,`ResponsePath`

`com.graphhopper.util`:

core: 
`AngleCalc`,`ArrayUtil`,`BitUtil/Big/Little`,`Breadth/Depth/XFirstSearch`,`DistanceCalc/3D/Earth/Euclidean/Projection`,`Constants`,`Downloader`,`EdgeExplorer/Iterator/State`,`FetchMode`,`GHUtility`,`GitInfo`,`Instructions`,`MiniPerfTest`,`PathMerger`,`PathSimplification`,`RamerDouglasPeucker`,`SimpleIntDeque`,`StopWatch`,`TranslationMap`,`Unzipper`

web-api: `CustomModel`,`FinishInstruction`,`Helper`,`Instruction`,`InstructionList`,`JsonFeature`,`JsonFeatureCollection`,`NumHelper`,`Parameters`,`PMap`,`PointAccess`,`PointList`,`RoundaboutInstruction`,`ShallowImmutablePointList`,`Translation`,`ViaInstruction`


`com.graphhopper.util.details`:

core:
`AbstractPathDetailsBuilder`,`EdgeIdDetails`,`KVStringDetails`,`TimeDetails`,`AverageSpeedDetails`,`EdgeKeyDetails`,`PathDetailsBuilderFactory`,`WeightDetails`,`BooleanDetails`,`EnumDetails`,`PathDetailsBuilder`,`DecimalDetails`,`IntDetails`,`PathDetailsFromEdges`,`DistanceDetails`,`IntersectionDetails`,`StringDetails`

web-api:
`PathDetail`

`com.graphhopper.util.shapes`:

core:
`BBox`,`Circle`,`Polygon`,`Shape`

web-api:
`GHPoint`,`GHPoint3D`


Is there still some tidying up to do with the classes or is it possible to rename the packages? I have made a pr that put  

- all the classes in `com.graphhopper` in the `core` module into `com.graphhopper.core` as well as
- all the classes in `com.graphhopper.util? in the `core` module into `com.graphhopper.core.util`.

That would fix the issue of split packages with the new Jigsaw module system and the next graphhopper release could be used with all maven projects with Java 9+.

Is it possible to finish this? It is now over 2,5 year ;).

Greetings from Germany

oli0044

**Comment by oli0044** [comment]
Heyhey :), 

any tendencies here?

In order to compare, I created a second pull request, showing what would be necessary to not change anything in `core` module but to do the changes in `web-api` module:

-     all the classes in `com.graphhopper` in the core module into `com.graphhopper.core` as well as
-     all the classes in `com.graphhopper.util` in the core module into `com.graphhopper.core.util`.

In summary: The amount of changes is similar. But it seems more logical and a bit less intrusive to move/rename the packages in `core` module.

What are your thoughts?

Greetings

Oli4

**Comment by easbar** [comment]
If we can choose between:

1) GraphHopper cannot be used with JDK9+ as soon as there is a module-info file somewhere in the project
2) All package names must be prefixed by the Maven module name (or some other, unique prefix)

I'd say we should go for 2) and we should do the package renaming before the 7.0 release. But maybe wait for a few other things to be merged to not create too much of a mess.

Do you have an example project that shows what exactly is currently not possible?

**Comment by oli0044** [comment]
Hi easbar,

I created a 1 file project in https://github.com/oli0044/graphhopper-java11-example (collaborator invite should also have been sent). 

It is not compiling because with the module-info.java, but compiling without it.
